fontir/src/orchestration.rs

```diff
@@ -34,83 +32,306 @@ bitflags! {
+    Ir: IrIo<I>,
+    T: Persistable,
+{
+    pub fn new(id: I, acl: Arc<AccessControlList<I>>, irio: Arc<Ir>) -> Self {
+        ContextItem {
+            id,
+            acl,
+            irio,
+            value: Default::default(),
+        }
+    }
+
+    pub fn clone_with_acl(&self, acl: Arc<AccessControlList<I>>) -> Self {
+        ContextItem {
+            id: self.id.clone(),
+            acl,
+            irio: self.irio.clone(),
+            value: self.value.clone(),
+        }
+    }
+
+    /// Read item that you are sure must exist. Panic if not.
+    ///
+    /// Intended for use in [Work] to access items present in
+    /// [Work::read_access]. If these are missing something is horribly
+    /// wrong and we should kerplode.
+    pub fn get(&self) -> Arc<T> {
+        self.acl.assert_read_access(&self.id);
+        if let Some(in_memory) = self.try_get() {
+            return in_memory;
+        }
+
+        // it's *not* in memory but perhaps it's written down?
+        if self.irio.emit_ir() {
+            if let Some(mut reader) = self.irio.reader(&self.id) {
+                let restored = T::read(&mut reader);
+                *self.value.write() = Some(Arc::from(restored));
+            }            
+        }
+
+        // if we still don't have an answer just give up
+        self.try_get().unwrap_or_else(|| panic!("{:?} is not available", self.id))
+    }
+
+    /// Read an item that might not exist
+    pub fn try_get(&self) -> Option<Arc<T>> {
+        self.acl.assert_read_access(&self.id);
+        self.value.read().as_ref().cloned()
+    }
+
+    /// Update the value of the item whether or not it has changed.
+    ///
+    /// The change will be logged and anything relying on change detection, such as
+    /// conditional execution of dependent tasks, will fire.
+    ///
+    /// [ContextItem::set] is preferable where possible.
+    ///
+    /// This exists largely because write types in fontations do not always implement PartialEq.
+    /// TODO: should they?
+    pub fn set_unconditionally(&self, value: T) {
+        self.acl.assert_write_access(&self.id);
+
+        if self.irio.emit_ir() {
+            let mut writer = self.irio.writer(&self.id);
+            value.write(&mut writer);
+        }
+
+        *self.value.write() = Some(Arc::from(value));
+    }
+}
+
+impl<I, T, Ir> ContextItem<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: PartialEq,
+    Ir: IrIo<I>,
+    T: Persistable,
+{
+    /// Update the value if it has changed.
+    ///
+    /// Change logging and dependent task execution will only fire if the value changed.
+    pub fn set(&self, value: T) {
+        self.acl.assert_write_access(&self.id);
+
+        // nop?
+        if self
+            .value
+            .read()
+            .as_ref()
+            .map(|arc| **arc == value)
+            .unwrap_or(false)
+        {
+            eprintln!("set_if_changed {:?}, NO CHANGE!", self.id);
```

Comment:
would logging be more appropriate?

Reply:
Oh taht should just go away. For context, I'm fond of using eprint for logs I want while playing that I don't intend to keep.